packages/providers/src/call-retrier.ts

```diff
@@ -1,70 +0,0 @@
-import type { ProviderOptions } from './provider';
-import { sleep } from './utils';
-
-type Backoff = 'linear' | 'exponential' | 'fixed';
-
-export interface RetryOptions {
-  /**
-   * Amount of attempts to retry before failing the call.
-   */
-  maxRetries: number;
-  /**
-   * Backoff strategy to use when retrying. Default is exponential.
-   */
-  backoff?: Backoff;
-  /**
-   * Base duration for backoff strategy. Default is 150ms.
-   */
-  baseDuration?: number;
-}
-
-function getWaitDuration(options: RetryOptions, attempt: number) {
-  const duration = options.baseDuration ?? 150;
-
-  if (attempt === 0) {
-    return duration;
-  }
-
-  switch (options.backoff) {
-    case 'linear':
-      return duration * attempt;
-    case 'fixed':
-      return duration;
-    case 'exponential':
-    default:
-      return duration * (2 ^ (attempt - 1));
-  }
-}
-
-export function retrier(
-  fetchFn: NonNullable<ProviderOptions['fetch']>,
-  options: RetryOptions | undefined,
-  retryAttempt: number = 0
-): NonNullable<ProviderOptions['fetch']> {
-  if (options === undefined) {
-    return fetchFn;
-  }
-
-  return async (...args) => {
-    try {
-      return await fetchFn(...args);
-    } catch (e: unknown) {
-      const error = e as Error & { cause?: { code?: string } };
-
-      if (error.cause?.code !== 'ECONNREFUSED') {
-        throw e;
-      }
```

Comment:
@nedsalk Note that I removed this `ECONNREFUSED` conditional.

Maybe I'm missing something, but I'd consider this an error that could _also_ be retried.

Reply:
The idea was to retry _only_ on failure to connect to the fuel node. There is no point in retrying on an error like "not enough coins...".

Reply:
I see, thanks for clarifying.

In this case, we must add new test cases to cover this.

How did you narrow this error down to only `ECONNREFUSED`?

Could there be other valid cases that are not accounted for? For example, what if the user gets a rate-limit response or something else?

Reply:
The discussions at the time were around not being able to connect so I just implemented that. I didn't think about other ways that the connection might be refused. I think it's best to create a discussion/issue around that.

Reply:
No worries, my main intention was to remove flakiness, which seems out of the way now. We can go like this now and revisit it later as necessary. Your implementation is almost as before, except for very tiny fixes, improvements, and docs. The big chunk of changes was by concentrating on unit-tests.

Note how it became easier to fully stress-test things in a unit fashion and keep an eye on the code coverage report to ensure you didn't forget to test anything after re-adding the `ECONNREFUSED` conditional.

```shell
pnpm test:filter packages/providers/src/utils/auto-retry-fetch.test.ts
open coverage/environments/node/index.html
```

![coverage](https://github.com/FuelLabs/fuels-ts/assets/26660/297cb882-d475-4bf7-b02d-c5dd828625e7)